### DISCUSSION TITLE
Re-trigger hash routing

### DISCUSSION THREAD

**Discussion by psinger** [h2oai/wave#703] [Q&A]
How can I solve the following scenario more conveniently for the user:

- A user clicks on navigation button with hash routing `#routing1`
- User navigates to http://localhost:10101/#routing1 which is also the current URL
- User operates on the new page, but taskbar stays the same as no new hash routing is triggered
- User clicks again on the same navigation button to reset the page / navigate back
- Hash routing is not triggered, as active URL is still the same

In simple terms, I would like to always trigger the hash routing, even if the same hash item is currently active.

**Comment by azim-b** [comment]
+1

This have been a shortcoming for a few use cases in multiple apps. Potential solution, if it is possible: "reset" the hash? (#1494)

**Comment by Far0n** [comment]
Possible workaround: Register an onclick events via js and use wave.emit:

```python
from h2o_wave import main, app, Q, ui


js = r"""
var links = document.links;
for(var i = 0; i < links.length; i++) {
    links[i].onclick = (e) => on_link_click(e, `hash`);
}
function on_link_click(e, val) {
    wave.debounce(50, () => { wave.emit(val, "clicked", true); })();
}
"""

@app("/demo")
async def serve(q: Q):
    print(f"q.args: {q.args}")
    print(f"q.events: {q.events}")
    print(f"q.client: {q.client}")
    
    # check q.events and avoid double clicking
    if "#" in q.args and not q.events.hash:
        q.client.new_hash = True
    elif "#" in q.args and q.events.hash and not q.client.new_hash:
        blurb = q.page["blurb"]
        blurb.content = f"Same hash clicked"
        q.client.new_hash = False
        return await q.page.save()
    else:
        q.client.new_hash = False

    if not q.client.initialized:
        q.client.initialized = True
        q.page["meta"] = ui.meta_card(box="", script=ui.InlineScript(content=js, targets=["a"]))

        q.page["nav"] = ui.markdown_card(
            box="1 1 4 2",
            title="Links!",
            content="[Spam](#menu/spam) / [Ham](#menu/ham) / [Eggs](#menu/eggs) / [About](#about)",
        )
        q.page["blurb"] = ui.markdown_card(
            box="1 3 4 2",
            title="Store",
            content="Welcome to our store!",
        )

    hash = q.args["#"]
    q.client.hash = hash
    blurb = q.page["blurb"]
    if hash == "menu/spam":
        blurb.content = f"Sorry, we're out of spam!"
    elif hash == "menu/ham":
        blurb.content = f"Sorry, we're out of ham!"
    elif hash == "menu/eggs":
        blurb.content = f"Sorry, we're out of eggs!"
    elif hash == "about":
        blurb.content = f"Everything here is gluten-free!"
    else:
        blurb.content = "Welcome to our store!"

    await q.page.save()

```



https://user-images.githubusercontent.com/6949295/173293517-e07e6d63-36d9-42ce-8feb-f72ad59e944c.mp4

**Comment by mturoci** [comment]
We might make it fire on every hash click (even when there is no hash change) unless there is not a strong reason why it was not implemented in this way in the first place - cc @lo5
